File: packages/git-core/src/plan/paths.ts
/**
 * Planners for path-scoped reads and writes.
 *
 * Path handling is where a Git client gets dangerous. Two rules are implemented
 * here and tested:
 *
 * 1. Paths are never interpolated into a shell string — they go into argv as
 *    separate elements, or into stdin as NUL-separated bytes.
 * 2. `--literal-pathspecs` is set whenever a path came from the user, so a file
 *    called `:(glob)**` is a file, not a pathspec that expands to everything.
 *    `--` separates paths from revisions, and the leading `-` problem is handled
 *    by the same literal mode plus `--`.
 */
import type { GitCommandSpec } from "../ports.js";
import type { PlanContext } from "./status.js";

function spec(
  context: PlanContext,
  argv: readonly string[],
  description: string,
  deadlineClass: GitCommandSpec["deadlineClass"] = "readonly",
  stdin?: Uint8Array,
): GitCommandSpec {
  return stdin === undefined
    ? { argv, cwdHandle: context.cwdHandle, deadlineClass, description }
    : { argv, cwdHandle: context.cwdHandle, deadlineClass, description, stdin };
}

/** NUL-join raw path bytes for a `--pathspec-from-file=… --pathspec-file-nul` stdin. */
export function pathspecStdin(paths: readonly Uint8Array[]): Uint8Array {
  if (paths.length === 0) {
    throw new Error("pathspecStdin requires at least one path");
  }
  let total = 0;
  for (const path of paths) {
    total += path.byteLength + 1;
  }
  const bytes = new Uint8Array(total);
  let offset = 0;
  for (const path of paths) {
    bytes.set(path, offset);
    offset += path.byteLength;
    bytes[offset] = 0x00;
    offset += 1;
  }
  return bytes;
}

/**
 * `git add` for exactly the selected paths.
 *
 * --literal-pathspecs: do not interpret `*`, `:(glob)`, etc. in our paths.
 * --pathspec-from-file with --pathspec-file-nul: the path list arrives as bytes on
 * stdin, so no path length or character limits apply and no quoting is needed.
 * --: ends options, so a path named `-n` cannot become a flag.
 */
export function planStage(
  context: PlanContext,
  paths: readonly Uint8Array[],
): GitCommandSpec {
  return spec(
    context,
    [
      "--literal-pathspecs",
      "add",
      "--pathspec-from-file=-",
      "--pathspec-file-nul",
      "--",
    ],
    "add paths",
    "hook",
    pathspecStdin(paths),
  );
}

/**
 * `git restore --staged` for the selected paths: index only, working tree
 * untouched. `--source=HEAD` is deliberately absent — unstage means "make the index
 * match HEAD", and when there is no HEAD the caller uses `planUnstageUnborn`.
 */
export function planUnstage(
  context: PlanContext,
  paths: readonly Uint8Array[],
): GitCommandSpec {
  return spec(
    context,
    [
      "--literal-pathspecs",
      "restore",
      "--staged",
      "--pathspec-from-file=-",
      "--pathspec-file-nul",
      "--",
    ],
    "restore --staged paths",
    "hook",
    pathspecStdin(paths),
  );
}

/**
 * Unstage in a repository with no commits: remove the entries from the index.
 *
 * `git restore --staged` needs a HEAD to restore from, and an unborn branch has
 * none, so this is a separate planner rather than a flag on the previous one —
 * the difference is exactly the kind of thing that silently produces an empty
 * repository if it is inferred.
 */
export function planUnstageUnborn(
  context: PlanContext,
  paths: readonly Uint8Array[],
): GitCommandSpec {
  return spec(
    context,
    [
      "--literal-pathspecs",
      "rm",
      "--cached",
      "-r",
      "--pathspec-from-file=-",
      "--pathspec-file-nul",
      "--",
    ],
    "rm --cached paths (unborn HEAD)",
    "hook",
    pathspecStdin(paths),
  );
}

/**
 * Restore working-tree files from the index.
 *
 * This is the "discard changes" primitive. It restores from the *index*, not from
 * HEAD — the design keeps those separate, and the caller is responsible for the
 * backup step before this runs.
 */
export function planRestoreWorktree(
  context: PlanContext,
  paths: readonly Uint8Array[],
): GitCommandSpec {
  return spec(
    context,
    [
      "--literal-pathspecs",
      "restore",
      "--worktree",
      "--pathspec-from-file=-",
      "--pathspec-file-nul",
      "--",
    ],
    "restore --worktree paths",
    "hook",
    pathspecStdin(paths),
  );
}

/** Changed paths between two revisions, with rename detection off (we report R/C). */
export function planDiffNameStatus(
  context: PlanContext,
  options: {
    readonly cached?: boolean;
    readonly from?: string;
    readonly to?: string;
  },
): GitCommandSpec {
  const argv = [
    "diff",
    "--name-status",
    "-z",
    "--no-ext-diff",
    "--no-textconv",
  ];
  if (options.cached === true) {
    argv.push("--cached");
  }
  if (options.from !== undefined) {
    argv.push(options.from);
  }
  if (options.to !== undefined) {
    argv.push(options.to);
  }
  return spec(context, argv, "diff --name-status -z");
}

export function planDiffNumstat(
  context: PlanContext,
  options: {
    readonly cached?: boolean;
    readonly from?: string;
    readonly to?: string;
  },
): GitCommandSpec {
  const argv = ["diff", "--numstat", "-z", "--no-ext-diff", "--no-textconv"];
  if (options.cached === true) {
    argv.push("--cached");
  }
  if (options.from !== undefined) {
    argv.push(options.from);
  }
  if (options.to !== undefined) {
    argv.push(options.to);
  }
  return spec(context, argv, "diff --numstat -z");
}

/**
 * Text patch for the whole change set (bounded by the caller's byte limit).
 *
 * `--submodule=short` makes a submodule change print its object names instead of
 * attempting to diff the submodule's contents.
 */
export function planDiffPatchAll(
  context: PlanContext,
  options: {
    readonly cached?: boolean;
    readonly from?: string;
    readonly to?: string;
  } = {},
): GitCommandSpec {
  const argv = [
    "diff",
    "--no-color",
    "--no-ext-diff",
    "--no-textconv",
    "--patch",
    "--submodule=short",
  ];
  if (options.cached === true) {
    argv.push("--cached");
  }
  if (options.from !== undefined) {
    argv.push(options.from);
  }
  if (options.to !== undefined) {
    argv.push(options.to);
  }
  return spec(context, argv, "diff patch (change set)");
}

/**
 * Text patch for exactly one path.
 *
 * The path is a string here because Git's argv is a string vector on this host;
 * the byte-level rule lives one level up: the host refuses a path whose bytes are
 * not representable, and the caller never substitutes a display form. The path is
 * also the *only* thing after `--`, and `--literal-pathspecs` keeps a file called
 * `:(glob)**` from expanding into every path in the repository.
 */
export function planDiffPatchForPath(
  context: PlanContext,
  options: {
    readonly path: string;
    readonly cached?: boolean;
    readonly from?: string;
    readonly to?: string;
  },
): GitCommandSpec {
  const argv = [
    "--literal-pathspecs",
    "diff",
    "--no-color",
    "--no-ext-diff",
    "--no-textconv",
    "--patch",
    "--submodule=short",
  ];
  if (options.cached === true) {
    argv.push("--cached");
  }
  if (options.from !== undefined) {
    argv.push(options.from);
  }
  if (options.to !== undefined) {
    argv.push(options.to);
  }
  argv.push("--", options.path);
  return spec(context, argv, "diff patch for one path");
}

/** Content of one blob, framed by the object protocol. Used for previews. */
export function planCatFileBlob(
  context: PlanContext,
  objectName: string,
): GitCommandSpec {
  const stdin = new Uint8Array(
    [...`${objectName}\n`].map((character) => character.charCodeAt(0)),
  );
  return spec(
    context,
    ["cat-file", "--batch"],
    "cat-file blob",
    "readonly",
    stdin,
  );
}

File: crates/refyard-core/src/plan/paths.rs
//! Planners for the path-scoped writes: stage and unstage.
//!
//! Two rules are implemented here, and both exist because of a way a Git client gets
//! dangerous:
//!
//! 1. **A path from the user is a path, not a pattern.** `--literal-pathspecs` is set on
//!    every command that receives one, so a file called `:(glob)**` or `*.txt` stages
//!    that file rather than expanding to a set nobody selected.
//! 2. **Paths travel as bytes on stdin, never interpolated into argv.** The list is
//!    NUL-delimited (`--pathspec-from-file=-` with `--pathspec-file-nul`), so a name
//!    containing a newline, a quote or a leading `-` cannot become a second argument or
//!    a flag — and `--` closes the option list even where a Git version ignores one of
//!    the two.
//!
//! Both are the strategies the Node reference's `planStage`/`planUnstage`/
//! `planUnstageUnborn` promise; the argument vectors here are the same, in the same
//! order, so the SSH transport sends the command the local transport runs.

use crate::problem::CoreError;

use super::{DeadlineClass, GitPlan};

/// NUL-joins raw path bytes for `--pathspec-from-file=- --pathspec-file-nul`.
///
/// A trailing NUL after every path, including the last: this is the format Git's
/// `--pathspec-file-nul` defines, and a last entry without one would be dropped by a Git
/// that only commits a pathspec at a terminator.
pub fn pathspec_stdin(paths: &[Vec<u8>]) -> Vec<u8> {
    let mut stdin = Vec::with_capacity(paths.iter().map(|path| path.len() + 1).sum());
    for path in paths {
        stdin.extend_from_slice(path);
        stdin.push(0);
    }
    stdin
}

/// The argument vector every path-scoped write shares: literal pathspecs, the path list
/// on stdin, NUL framing, and `--` before anything could be read as an option.
fn pathspec_plan(verb: &[&str], paths: &[Vec<u8>]) -> GitPlan {
    let mut argv: Vec<String> = vec!["--literal-pathspecs".to_string()];
    argv.extend(verb.iter().map(|value| (*value).to_string()));
    argv.push("--pathspec-from-file=-".to_string());
    argv.push("--pathspec-file-nul".to_string());
    argv.push("--".to_string());
    GitPlan {
        argv,
        stdin: pathspec_stdin(paths),
        // Staging runs the user's hooks, filters and signer, so it gets the hook
        // deadline rather than a read's.
        deadline_class: DeadlineClass::Hook,
    }
}

/// `git add` for exactly the selected paths.
///
/// Nothing here stages implicitly: the paths named are the paths staged, and the index
/// entries for every other changed path are left as they are.
pub fn plan_stage(paths: &[Vec<u8>]) -> GitPlan {
    pathspec_plan(&["add"], paths)
}

/// `git restore --staged` for the selected paths: the index only, working tree untouched.
///
/// `--source=HEAD` is deliberately absent. Unstage means "make the index match HEAD", and
/// spelling the source would make this plan wrong for a repository whose HEAD is a ref
/// that has not been created yet; that case is [`plan_unstage_unborn`].
pub fn plan_unstage(paths: &[Vec<u8>]) -> GitPlan {
    pathspec_plan(&["restore", "--staged"], paths)
}

/// Unstage in a repository with no commits: remove the entries from the index.
///
/// `git restore --staged` needs a HEAD to restore from, and an unborn branch has none, so
/// this is a separate plan rather than a flag on the previous one: the difference is
/// exactly the kind of thing that silently leaves an empty repository if it is inferred.
/// It removes index entries only — the working-tree files are not touched.
pub fn plan_unstage_unborn(paths: &[Vec<u8>]) -> GitPlan {
    pathspec_plan(&["rm", "--cached", "-r"], paths)
}

/// Refuses a path list that would produce a command with nothing to do.
///
/// `git add` with an empty pathspec list stages everything under the current directory,
/// which is the opposite of "the paths the user selected". Every caller runs this before
/// planning, so an empty selection is a typed refusal rather than a whole-repository add.
pub fn require_paths(paths: &[Vec<u8>]) -> Result<(), CoreError> {
    if paths.is_empty() {
        return Err(CoreError::invalid_input(
            "a path-scoped write needs at least one path",
        ));
    }
    for path in paths {
        if path.contains(&0) {
            return Err(CoreError::unsupported_path_encoding(
                "pathspec",
                "a path containing NUL cannot address a file on any POSIX host",
            ));
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn paths(values: &[&str]) -> Vec<Vec<u8>> {
        values
            .iter()
            .map(|value| value.as_bytes().to_vec())
            .collect()
    }

    #[test]
    fn stage_passes_literal_pathspecs_on_stdin_with_nul_framing() {
        // Prevents: `git add` reading a file called `:(glob)**` as a pattern, and a path
        // containing a newline splitting into two pathspecs.
        let plan = plan_stage(&paths(&["a.txt", ":(glob)**", "line\nbreak.txt"]));
        assert_eq!(
            plan.argv,
            vec![
                "--literal-pathspecs",
                "add",
                "--pathspec-from-file=-",
                "--pathspec-file-nul",
                "--"
            ]
        );
        assert_eq!(
            plan.stdin,
            b"a.txt\0:(glob)**\0line\nbreak.txt\0".to_vec(),
            "the path list is bytes on stdin, so no quoting or length rule applies"
        );
        assert_eq!(plan.deadline_class, DeadlineClass::Hook);
    }

    #[test]
    fn unstage_restores_the_index_only_and_unborn_removes_the_entries() {
        // The two plans differ by design; inferring one from the other is how an unborn
        // repository ends up with its files deleted or its index untouched.
        let staged = plan_unstage(&paths(&["a.txt"]));
        assert_eq!(staged.argv[1], "restore");
        assert_eq!(staged.argv[2], "--staged");
        assert!(
            !staged.argv.contains(&"--source=HEAD".to_string()),
            "an explicit source is what an unborn repository cannot answer"
        );

        let unborn = plan_unstage_unborn(&paths(&["a.txt"]));
        assert_eq!(
            unborn.argv,
            vec![
                "--literal-pathspecs",
                "rm",
                "--cached",
                "-r",
                "--pathspec-from-file=-",
                "--pathspec-file-nul",
                "--"
            ],
            "`rm --cached` removes index entries and never touches the working tree"
        );
        assert_eq!(unborn.stdin, b"a.txt\0".to_vec());
    }

    #[test]
    fn an_empty_selection_or_a_nul_in_a_path_is_refused_before_a_command_exists() {
        // An empty path list is an `add` of everything, which is not what the caller
        // asked for; NUL cannot travel through a pathspec on stdin.
        assert!(require_paths(&[]).is_err());
        assert!(require_paths(&[b"bad\0name".to_vec()]).is_err());
        assert!(require_paths(&paths(&["a.txt"])).is_ok());
    }
}
